lib/rspec/matchers/built_in/compound.rb

```diff
@@ -155,6 +155,13 @@ def initialize(actual, matcher_1, matcher_2)
 
           def matcher_matches?(matcher)
             @match_results.fetch(matcher)
+          rescue IndexError
```

Comment:
Instead of rescuing this, I believe you can just pass this `raise` expression in a block to `fetch`.  The block is called only if the provided key is absent.  It's an explicit, built-in way with `fetch` to run custom logic when the key is missing.